pdns/pdns_recursor.cc

```diff
@@ -396,114 +397,132 @@ static bool isHandlerThread()
+  TCPLOG("asendtcp waitEvent returned " << ret << ' ' << packet.size() << '/' << data.size() << ' ');
+  if (ret == 0) {
+    TCPLOG("timeout" << endl);
+    TCPIOHandlerStateChange(pident->lowState, IOState::Done, pident);
     return LWResult::Result::Timeout;
   }
   else if (ret == -1) { // error
-    t_fdm->removeWriteFD(handler.getDescriptor());
+    TCPLOG("PermanentError" << endl);
+    TCPIOHandlerStateChange(pident->lowState, IOState::Done, pident);
     return LWResult::Result::PermanentError;
   }
   else if (packet.size() != data.size()) { // main loop tells us what it sent out, or empty in case of an error
+    // fd housekeeping done by TCPIOHandlerIO
+    TCPLOG("PermanentError size mismatch" << endl);
     return LWResult::Result::PermanentError;
   }
 
+  TCPLOG("asendtcp success" << endl);
   return LWResult::Result::Success;
 }
 
-static void handleTCPClientReadable(int fd, FDMultiplexer::funcparam_t& var);
-
-LWResult::Result arecvtcp(PacketBuffer& data, const size_t len, Socket* sock, const bool incompleteOkay)
+LWResult::Result arecvtcp(PacketBuffer& data, const size_t len, shared_ptr<TCPIOHandler>& handler, const bool incompleteOkay)
 {
-  data.clear();
-  PacketID pident;
-  pident.tcpsock=sock->getHandle();
-  pident.inNeeded=len;
-  pident.inIncompleteOkay=incompleteOkay;
-  t_fdm->addReadFD(sock->getHandle(), handleTCPClientReadable, pident);
+  TCPLOG("arecvtcp called " << len << ' ' << data.size() << endl);
+  data.resize(len);
 
-  int ret = MT->waitEvent(pident, &data, g_networkTimeoutMsec);
-  if (ret == 0) {
-    t_fdm->removeReadFD(sock->getHandle());
-    return LWResult::Result::Timeout;
-  }
-  else if (ret == -1) {
-    t_fdm->removeWriteFD(sock->getHandle());
-    return LWResult::Result::PermanentError;
+  // We might have data already available from the TLS layer, try to get that into the buffer
+  size_t pos = 0;
+  IOState state;
+  try {
+    TCPLOG("calling tryRead() " << len << endl);
+    state = handler->tryRead(data, pos, len);
+    TCPLOG("arcvtcp tryRead() returned " << int(state) << ' ' << pos << '/' << len << endl);
+    switch (state) {
+    case IOState::Done:
+    case IOState::NeedRead:
+      if (pos == len || (incompleteOkay && pos > 0)) {
+        data.resize(pos);
+        TCPLOG("acecvtcp success A" << endl);
+        return LWResult::Result::Success;
+      }
+      break;
+    case IOState::NeedWrite:
+      break;
+    }
   }
-  else if (data.empty()) {// error, EOF or other
+  catch (const std::exception& e) {
+    TCPLOG("tryRead() exception..." << e.what() << endl);
     return LWResult::Result::PermanentError;
   }
 
-  return LWResult::Result::Success;
-}
-
-static void TCPIOHandlerReadable(int fd, FDMultiplexer::funcparam_t& var);
+  auto pident = std::make_shared<PacketID>();
+  pident->tcphandler = handler;
+  pident->tcpsock = handler->getDescriptor();
+  // We might have a partial result
+  pident->inMSG = data;
```

Comment:
Since we are going to clear it below anyway:
```suggestion
  pident->inMSG = std::move(data);
```